lib/src/network/protocol/proxy_protocol.rs

```diff
@@ -0,0 +1,121 @@
+use std::net::IpAddr;
+use std::io::{Write, ErrorKind};
+use mio::*;
+use mio::tcp::TcpStream;
+use mio::unix::UnixReady;
+use network::{ClientResult, Protocol};
+use network::socket::{SocketHandler,SocketResult};
+use network::protocol::ProtocolResult;
+use network::session::Readiness;
+use network::header_proxy_protocol::*;
+use network::buffer_queue::BufferQueue;
+use pool::Checkout;
+
+pub struct ProxyProtocol<Front:SocketHandler> {
+  pub header:     Option<ProxyProtocolHeader>,
+  pub frontend:   Front,
+  pub backend:    Option<TcpStream>,
+  frontend_token: Option<Token>,
+  backend_token:  Option<Token>,
+  pub readiness:  Readiness,
+  cursor_header:  usize,
+  pub front_buf:      Checkout<BufferQueue>,
+  pub back_buf:       Checkout<BufferQueue>,
+}
+
+impl <Front:SocketHandler>ProxyProtocol<Front> {
+  pub fn new(frontend: Front, backend: Option<TcpStream>, front_buf: Checkout<BufferQueue>,
+    back_buf: Checkout<BufferQueue>) -> Self {
+
+    ProxyProtocol {
+      header: None,
+      frontend,
+      backend,
+      frontend_token: None,
+      backend_token:  None,
+      readiness: Readiness {
+        front_interest:  UnixReady::hup() | UnixReady::error(),
+        back_interest:   UnixReady::from(Ready::writable()) | UnixReady::hup() | UnixReady::error(),
+        front_readiness: UnixReady::from(Ready::empty()),
+        back_readiness:  UnixReady::from(Ready::empty()),
+      },
+      cursor_header: 0,
+      front_buf,
+      back_buf,
+    }
+  }
+
+  // The header is send immediately at once upon the connection is establish
+  // and prepended before any data.
+  pub fn back_writable(&mut self) -> (ProtocolResult, ClientResult) {
+    debug!("Writte proxy protocol header");
+
+    if let Some(ref mut socket) = self.backend {
+      if let Some(ref header) = self.header {
+        let mut header = header.into_bytes();
+        loop {
+          match socket.write(&mut header[self.cursor_header..]) {
+            Ok(sz) => {
+              self.cursor_header += sz;
+
+              if self.cursor_header == header.len() {
+                debug!("The header should be receive. We switch to a Pipe");
```

Comment:
"proxy protocol sent, upgrading" (not specific to TCP, and we do not know if the header was received, we only know we sent it)

Reply:
Ok

Reply:
Done